### PR TITLE
Copter: set EKF origin from first do-set-home command

### PR DESCRIPTION
**Pull request by rmackay9**
This PR allows the EKF origin to be set from the first do-set-home command (received from the GCS) if it arrives before the EKF has set it's origin.

This is important for cases where the vehicle only has a relative velocity or position source like visual odometry or optical flow.  Without an origin, navigation commands that require an absolute position target (i.e. takeoff, waypoint, spline, etc) will fail.  This means that currently AUTO and GUIDED flight modes do not work with only visual odometry or optical flow. 

There's a little bit of unnecessary code in the AP_AHRS's set_origin call related to whether the method returns true or false.  In fact, we currently never use this return value anyway so we could remove this although I think what it returns is reasonable.

### CODE REVIEW

**Comment by rmackay9** [comment]
By the way, I have not implemented this for Sub, Plane, Rover or Tracker but we easily could.

File: libraries/AP_AHRS/AP_AHRS.h [hunk 1]
```diff
@@ -441,6 +441,11 @@ class AP_AHRS
     // current barometer and GPS altitudes correspond to this altitude
     virtual void set_home(const Location &loc) = 0;
 
+    // set the EKF's origin location in 10e7 degrees.  This should only
+    // be called when the EKF has no absolute position reference (i.e. GPS)
+    // from which to decide the origin on its own
+    virtual bool set_origin(const Location &loc) { return true; }
```
Comment (OXINARF) [inline libraries/AP_AHRS/AP_AHRS.h]: I think returning false by default would be better, but we aren't actually doing anything with it.

**Comment by rmackay9** [comment]
Here's an attempt at a logic flow diagram for how the AHRS home and EKF origin are set.  Not a great diagram but I hope it helps a bit.
homesetlogic

**Comment by rmackay9** [comment]
Thanks for your review and comments.

From a light investigation, AC_WPNav methods that fails (for guided takeoff) is one that accepts a Location_Class parameter as the target destination.  I think it fails because it attempts to convert the Location into an offset from the EKF origin which is what the navigation and position control libraries use internally.

I'll need to look into why the RTL actually works.  I suspect that it's assuming the ahrs home and ekf origin are at 0,0,0 (i.e. off the west coast of Africa) and/or it's using the AC_WPNav methods that accept Vector3f as targets which are assumed to be offsets from the EKF origin.

It's possible we're building on shaky foundations but I suspect this PR is still an OK approach because it should be possible for vehicles to fly in locations where they don't have a GPS but the user is willing to specify the vehicle's starting location.

**Comment by OXINARF** [comment]
Thanks for looking into it. I think that making sure we can use relative frame positions is a worthwhile job to pursue.

I wasn't clear before, but I fully support this PR. I think it can be helpful for someone not using a GPS (or it taking a while to get a good signal) and still be able to use absolute positions.
I think we should limit when origin can be set though - I'm not sure setting it in the air is a safe code path. @priseborough should give his input on this.
Also, if you haven't talked with Paul about it, we should ask him how he thinks this might interact in the EKF if a GPS becomes active and there is a big shift in position. We can also limit the origin setting only for when there is no GPS attached (it may be restricting too much though).

**Comment by cglusky** [comment]
Use Case to consider to support PR:
- Urban or natural canyon with no/min GPS performance
- Takeoff with only VOD or similar sensor to get above canyon walls
- Switch to GPS when able
- Reverse process for landing

This can also be an issue around large vessels that can cause RF and/or mag interference.

Not sure this is an edge case.  Not enough data on my end to make any concrete claims.

**Comment by WickedShell** [comment]
I recall conversations with @priseborough about the dangers of calling setOriginLLH() on the EKF, but it might be absolutely correct in this case?

**Comment by rmackay9** [comment]
Paul hasn't seen these exact changes yet but I implemented them after talking with him about it.  In any case, I ping'd him just now and I think he'll add some comments in the near-ish future.

**Comment by priseborough** [comment]
I am comfortable with setting an EKF origin at EKF if one is not already set as this will enable useful WGS-84 estimates to be obtained from the EKF even when it has no GPS and is in relative position mode. 

As long as the EKF is operating in relative position mode using either optical flow or odometry data, then subsequent changes to the origin lat/long should not effect the EKF itself because the origin will only be used to convert internal local NE position states to an absolute lat/long value for output. This will however make the WGS-84 position reported by getLLH will jump  which will affect navigation loops depending on the navigation mode.  Likewise for subsequent changes to origin height. Provided baro or range finder height is being used, the internal local down position state will be unaffected, however the height reported by getLLH will jump.

File: libraries/AP_AHRS/AP_AHRS.h (hunk 1 above)
Comment (rmackay9) [inline libraries/AP_AHRS/AP_AHRS.h]: ok fixed to return false.

**Comment by rmackay9** [comment]
I've updated it so that the EKF origin can only be set when the vehicle is disarmed.  We don't know that it will cause problems to change while in flight but it might so for now at least, this is the safe approach.

**Comment by rmackay9** [comment]
Thanks very much for the review!

**Pull request merged**